Revision v1.2

chuan-wang opened the pull request:
1) Add support to use the miRBase sequences of ALL species as reference;
2) Wrap result files for 'mature' and 'hairpin' in different publish directories;
3) Small changes to add and remove some result files.

chuan-wang committed:
Revision v1.1f
Small corrections on the -p parameter for bowtie2 alignment;
Remove redundant script for prefix in miRBasePostAlignment

chuan-wang committed:
Revision v1.2
1) Add support to use the miRBase sequences of ALL species as reference;
2) Wrap result files for 'mature' and 'hairpin' in different publish directories;
3) Small changes to add and remove some result files.

chuan-wang committed:
Revision v1.2
Add instruction to use the miRBase sequences of ALL species as reference

ewels commented on main.nf:
```diff
@@ -209,7 +228,7 @@ process bowtie_miRBase_mature {
     maxErrors '-1'
 
-    publishDir "${params.outdir}/bowtie_miRBase_mature", mode: 'copy'
+    publishDir "${params.outdir}/bowtie_miRBase_mature", mode: 'copy', pattern: '*.mature_unmapped.fq.gz'
```
Is there a reason for not delivering BAM files?

chuan-wang replied:
We deliver the sorted bam file in the miRBasePostAlignment step.

ewels commented on main.nf:
```diff
@@ -295,6 +314,14 @@ process bowtie_miRBase_hairpin {
 }
 
 
+def rule_step5(file){
```
Can we have a nicer name for this? `def mature_hairpin_publishDirs` or something. `rule_step5` is a bit non-descriptive, and the step numbers are easily changed.

chuan-wang replied:
Changed.

ewels commented on main.nf:
```diff
@@ -311,17 +338,19 @@ process miRBasePostAlignment {
     maxErrors '-1'
 
-    publishDir "${params.outdir}/miRBaseCounts", mode: 'copy'
+    publishDir "${params.out}", mode: 'copy', saveAs: this.&rule_step5
```
`${params.out}` not `${params.outdir}`? Is this actually used at all? If not might be nice to make that clear:

``` groovy
publishDir 'set_by_rule', mode: 'copy', saveAs: this.&rule_step5
```

chuan-wang replied:
`${params.outdir}` is the `./results` folder. It will create one `results` folder under the `./results` folder. That's why we use the `${params.out}` instead.

ewels replied:
Ok, so this is passed to the function? In that case, can we pass `$params.outdir` and then append to that in the function call instead.. I don't see why we need `$params.out` as well? In fact, that variable doesn't seem to be used by anything else in the script, so can we remove it entirely?

That wasn't very clear. Proposed code:

``` groovy
publishDir $params.outdir, mode: 'copy', saveAs: this.&rule_step5
```

``` groovy
def rule_step5(file){
      if ( file.contains("mature") )
        return "/bowtie_miRBase_mature/$file"

      if ( file.contains("hairpin") )
        return "/bowtie_miRBase_hairpin/$file"
}
```

Better still, we could have a single function for both this process _and_ the edgeR process:

``` groovy
publishDir "${params.outdir}/bowtie_miRBase", mode: 'copy', saveAs: this.&mature_hairpin_dirs
```

``` groovy
publishDir "${params.outdir}/edgeR", mode: 'copy', saveAs: this.&mature_hairpin_dirs
```

``` groovy
def mature_hairpin_dirs(file){
      if ( file.contains("mature") )
        return "/mature/$file"

      if ( file.contains("hairpin") )
        return "/hairpin/$file"
}
```

chuan-wang replied:
Combined as one function in v1.2b

ewels replied:
Great! Does it work?

chuan-wang replied:
Yes, I've tested the entire pipeline this time. Still waiting for results from multiqc though but it seems to be working at least

ewels commented on main.nf:
```diff
@@ -330,6 +359,14 @@ process miRBasePostAlignment {
 }
 
 
+def rule_step6(file){
```
Maybe a nicer name again as above.

Could make an argument to try to use a single function for both, but it's not many lines and I think this is clearer 👍

ewels commented on main.nf:
```diff
@@ -346,7 +383,7 @@ process edgeR_miRBase_mature {
     maxErrors '-1'
 
-    publishDir "${params.outdir}/edgeR", mode: 'copy'
+    publishDir "${params.out}", mode: 'copy', saveAs: this.&rule_step6
```
As above with `${params.out}`

chuan-wang replied:
Specified above.

ewels commented on main.nf:
```diff
@@ -425,19 +462,30 @@ process edgeR_miRBase_mature {
+
     # Print distance matrix to file
-    write.table(MDSdata\$distance.matrix, paste(header,"_edgeR_MDS_distance_matrix.txt",sep=""), quote=FALSE, sep="\\t")
+    # write.table(MDSdata\$distance.matrix, paste(header,"_edgeR_MDS_distance_matrix.txt",sep=""), quote=FALSE, sep="\\t")
```
Is this meant to be commented out? If it's not being used, delete it instead of commenting it. We can always get it back again using version control.

chuan-wang replied:
Deleted.

ewels replied:
I forgot to ask.. Why are you commenting out / deleting these files? The heatmap and MDS plot in the MultiQC report will disappear without them..

chuan-wang replied:
Aha...I deleted them in order to minimize the output files from edgeR. I will add them back then...but they will not be published as results of edgeR.

ewels replied:
Ok, that's fine.

ewels commented on main.nf:
```diff
@@ -569,7 +626,6 @@ process multiqc {
 
     output:
     file '*multiqc_report.html'
-    file '*multiqc_data'
```
Why are you removing the parsed data directory?

chuan-wang replied:
I'm thinking that we don't really need to deliver the parsed data files to the user. They can find all info from the html file. Many raw files may cause our trouble to explain everything...

ewels replied:
No, I disagree sorry. The `multiqc_data` directory contains a lot of information which is not present in the HTML report - MultiQC generally parses all of the numbers it can and only puts some of the key values into the HTML report. The data directory contains a lot of other stats. It's also fairly easy to explain I think..?

chuan-wang replied:
OK, then I will put them back.

ewels reviewed (changes requested):
Great! A few minor comments but all nice changes.

chuan-wang committed:
Revision v1.2b
1) Adjusted the publish directories for bowtie output files;
2) Modified the wrap function for dividing mature and hairpin output files into different directories;
3) Removed some output files from edgeR;
4) Slightly changed the scripts for multiqc.

chuan-wang committed:
Revision v1.2c
Restored the scripts for generating data files in edgeR.
